src/NumericalAlgorithms/Integration/GslQuadAdaptive.hpp

```diff
@@ -0,0 +1,261 @@
+    p | max_intervals_;
+    if (p.isUnpacking()) {
+      initialize();
+    }
+  }
+
+  double (*function_)(double, Args...);
+  std::tuple<Args...> args_;
+
+  template <size_t Index, typename Arg>
+  void set_parameter(const Arg& arg) {
+    std::get<Index>(args_) = arg;
+  }
+
+  template <typename IntegrandType>
+  void set_integrand(IntegrandType&& integrand) {
+    function_ = integrand;
+  }
+
+ private:
+  struct gsl_integration_workspace_deleter {
+    void operator()(gsl_integration_workspace* workspace) const noexcept {
+      gsl_integration_workspace_free(workspace);
+    }
+  };
+
+  void initialize() noexcept {
+    workspace_ = std::unique_ptr<gsl_integration_workspace,
+                                 gsl_integration_workspace_deleter>{
+        gsl_integration_workspace_alloc(max_intervals_)};
+    integrand_.function = &detail::integrand<Args...>;
+    integrand_.params = this;
+  }
+
+ protected:
+  size_t max_intervals_;
+  gsl_function integrand_;
+
+  std::unique_ptr<gsl_integration_workspace, gsl_integration_workspace_deleter>
+      workspace_;
+};
+}  // namespace detail
+
+/*!
+ * \ingroup NumericalAlgorithmsGroup
+ * \brief A adaptive integration procedure class
+ *
+ * The class builds a numerical integrator with either an absolute error or
+ * relative error and the maximum number of subintervals passed into the
+ * constructor. For details on the algorithm see the GSL documentation on
+ * `gsl_integration`.
+ *
+ * Here is an example how to use this class:
+ * For the function:
+ * \snippet Test_GslQuad.cpp integrated_function
+ * the integration should look like:
+ * \snippet Test_GslQuad.cpp integration_example
+ */
+
+enum class IntegralType {
+  StandardGaussKronrod,            // gsl_integration_qag()
+  IntegrableSingularitiesPresent,  // gsl_integration_qags()
+  IntegrableSingularitiesKnown,  // gsl_integration_qagp(), domain given by pts
+  InfiniteInterval,              // gsl_integration_qagi()
+  UpperBoundaryInfinite,         // gsl_integration_qagiu()
+  LowerBoundaryInfinite          // gsl_integration_qagil()
+};
+
+template <IntegralType TheIntegralType, typename... Args>
+class GslQuadAdaptive;
+
+template <typename... Args>
+class GslQuadAdaptive<IntegralType::StandardGaussKronrod, Args...>
+    : public detail::GslQuadAdaptiveImpl<Args...> {
+ public:
+  using detail::GslQuadAdaptiveImpl<Args...>::GslQuadAdaptiveImpl;
+  // gsl_integration_qag(), applies Gauss-Kronrod rules with key as index
+  // to no_points = [15, 21, 31, 41, 51, 61]
+  Result operator()(double lower_boundary, double upper_boundary,
+                    double tolerance_abs, int key) const noexcept {
+    Result result;
+    int status = gsl_integration_qag(
+        &(this->integrand_), lower_boundary, upper_boundary, tolerance_abs, 0.,
+        this->max_intervals_, key, this->workspace_.get(), &result.value,
+        &result.error);
+    if (status != 0.) {
+      ::Parallel::printf("error: %s\n", gsl_strerror(status));
```

Comment:
Please remove all of these `printf`s.

We have a few options how to handle the `status` and the `error`:
- Add another enum for the possible status values and add it to the `Result` type
- Directly return the result as a double and use member variables for the `error` and the `status` instead of the `Result` type

Perhaps whoever does the second review could comment on this. You can leave the code as-is until then (but remove the prints).